File: LocationReservation.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#define MAX_LOCATIONS 100
#define MAX_RESERVATIONS 100

typedef enum { VEHICLE, PROPERTY } LocationType;

typedef struct {
    int id;
    char name[100];
    LocationType type;
    int available;
} Location;

typedef struct {
    int locationId;
    struct tm startDate;
    struct tm endDate;
    int durationInDays;
} Reservation;

Location locations[MAX_LOCATIONS];
Reservation reservations[MAX_RESERVATIONS];
int locationCount = 0, reservationCount = 0;

void addLocation(int id, const char *name, LocationType type) {
    if (locationCount < MAX_LOCATIONS) {
        locations[locationCount++] = (Location){id, "", type, 1};
        strcpy(locations[locationCount - 1].name, name);
    }
}

void listLocations() {
    for (int i = 0; i < locationCount; i++) {
        printf("ID: %d, Name: %s, Available: %s\n", locations[i].id, locations[i].name, locations[i].available ? "Yes" : "No");
    }
}

int isReserved(int locationId, struct tm startDate, struct tm endDate) {
    for (int i = 0; i < reservationCount; i++) {
        if (reservations[i].locationId == locationId) {
            if (difftime(mktime(&endDate), mktime(&reservations[i].startDate)) >= 0 &&
                difftime(mktime(&startDate), mktime(&reservations[i].endDate)) <= 0) {
                return 1;
            }
        }
    }
    return 0;
}

void createReservation(int locationId, struct tm startDate, struct tm endDate) {
    for (int i = 0; i < locationCount; i++) {
        if (locations[i].id == locationId && locations[i].available) {
            if (isReserved(locationId, startDate, endDate)) {
                printf("Error: Location already reserved.\n");
                return;
            }

            int duration = (int)difftime(mktime(&endDate), mktime(&startDate)) / (60 * 60 * 24);
            if (duration < 1) {
                printf("Error: Duration must be at least 1 day.\n");
                return;
            }

            reservations[reservationCount++] = (Reservation){locationId, startDate, endDate, duration};
            printf("Reservation confirmed for %s for %d days.\n", locations[i].name, duration);
            return;
        }
    }
    printf("Error: Location not found or unavailable.\n");
}

int main() {
    addLocation(1, "Toyota Corolla", VEHICLE);
    addLocation(2, "Appartement T3", PROPERTY);

    printf("--- Location Manager CLI ---\n");
    listLocations();

    struct tm startDate = {0}, endDate = {0};
    startDate.tm_year = 2025 - 1900; startDate.tm_mon = 4 - 1; startDate.tm_mday = 10;
    endDate.tm_year = 2025 - 1900; endDate.tm_mon = 4 - 1; endDate.tm_mday = 12;

    createReservation(1, startDate, endDate);

    return 0;
}

File: LocationReservation.rs

use std::collections::HashMap;
use std::io::{self, Write};
use chrono::{NaiveDate, Duration};

#[derive(Debug, Clone, PartialEq)]
enum LocationType {
    Vehicle,
    Property,
}

#[derive(Debug, Clone)]
struct Location {
    id: u32,
    name: String,
    location_type: LocationType,
    available: bool,
}

#[derive(Debug, Clone)]
struct Reservation {
    location_id: u32,
    start_date: NaiveDate,
    end_date: NaiveDate,
    duration: i64,
}

struct LocationManager {
    locations: HashMap<u32, Location>,
    reservations: Vec<Reservation>,
}

impl LocationManager {
    fn new() -> Self {
        Self { locations: HashMap::new(), reservations: Vec::new() }
    }

    fn add_location(&mut self, id: u32, name: &str, location_type: LocationType) {
        self.locations.insert(id, Location {
            id,
            name: name.to_string(),
            location_type,
            available: true,
        });
    }

    fn list_locations(&self) {
        for location in self.locations.values() {
            println!("ID: {}, Name: {}, Available: {}", location.id, location.name, location.available);
        }
    }

    fn create_reservation(&mut self, location_id: u32, start_date: NaiveDate, end_date: NaiveDate) -> Result<(), String> {
        if let Some(location) = self.locations.get(&location_id) {
            if !location.available { return Err("Location not available.".to_string()); }

            for res in &self.reservations {
                if res.location_id == location_id &&
                   (start_date <= res.end_date && end_date >= res.start_date) {
                    return Err("Location already reserved in this period.".to_string());
                }
            }

            let duration = (end_date - start_date).num_days();
            if duration < 1 { return Err("Duration must be at least 1 day.".to_string()); }

            self.reservations.push(Reservation { location_id, start_date, end_date, duration });
            Ok(())
        } else {
            Err("Location not found.".to_string())
        }
    }
}

fn main() {
    let mut manager = LocationManager::new();

    manager.add_location(1, "Toyota Corolla", LocationType::Vehicle);
    manager.add_location(2, "Appartement T3", LocationType::Property);

    println!("--- Location Manager CLI ---");
    manager.list_locations();

    let start_date = NaiveDate::from_ymd_opt(2025, 4, 10).unwrap();
    let end_date = NaiveDate::from_ymd_opt(2025, 4, 12).unwrap();

    match manager.create_reservation(1, start_date, end_date) {
        Ok(_) => println!("Reservation confirmed."),
        Err(e) => println!("Error: {}", e),
    }
}
